lib/charms/vault_k8s/v0/vault_managers.py

```diff
@@ -504,3 +526,288 @@ def generate_vault_unit_certificate(
+            for relation in self._juju_facade.get_active_relations(self._provides.relation_name)
+        ]
+        for policy in existing_policies:
+            if policy not in relation_policy_names:
+                self._client.delete_policy(policy)
+                logging.info("Removing unused policy: %s", policy)
+
+    def _create_key(self, key_name) -> None:
+        response = self._client.create_transit_key(mount_point=self.mount_path, key_name=key_name)
+        logging.debug("Created a new autounseal key: %s", response)
+
+    def create_credentials(self, relation: Relation) -> tuple[str, str, str]:
+        """Create auto-unseal credentials for the given relation.
+
+        Args:
+            relation: The relation to create the credentials for.
+
+        Returns:
+            A tuple containing the key name, role ID, and secret ID.
+        """
+        relation_details = VaultAutounsealRelationDetails(relation)
+        self._create_key(relation_details.key_name)
+        policy_content = AUTOUNSEAL_POLICY.format(
+            mount=self.mount_path, key_name=relation_details.key_name
+        )
+        self._client.create_or_update_policy(
+            relation_details.policy_name,
+            policy_content,
+        )
+        role_id = self._client.create_or_update_approle(
+            relation_details.approle_name,
+            policies=[relation_details.policy_name],
+            token_period="60s",
+        )
+        secret_id = self._client.generate_role_secret_id(relation_details.approle_name)
+        self._provides.set_autounseal_data(
+            relation,
+            self.get_address(relation),
+            self.mount_path,
+            relation_details.key_name,
+            role_id,
+            secret_id,
+            self._ca_cert,
+        )
+        return relation_details.key_name, role_id, secret_id
+
+    def _get_existing_keys(self) -> list[str]:
+        return self._client.list(f"{self.mount_path}/keys")
+
+    def _get_existing_roles(self) -> list[str]:
+        output = self._client.list("auth/approle/role")
+        return [role for role in output if role.startswith("charm-autounseal-")]
+
+    def _get_existing_policies(self) -> list[str]:
+        output = self._client.list("sys/policy")
+        return [policy for policy in output if policy.startswith("charm-autounseal-")]
+
+
+@dataclass
+class AutounsealConfigurationDetails:
+    """Credentials required for configuring auto-unseal on Vault."""
+
+    address: str
+    mount_path: str
+    key_name: str
+    token: str
+    ca_cert_path: str
+
+
+class VaultAutounsealRequirerManager:
+    """Encapsulates the auto-unseal functionality from the Requirer Perspective.
+
+    In other words, this manages the feature from the perspective of the Vault
+    being auto-unsealed.
+    """
+
+    AUTOUNSEAL_TOKEN_SECRET_LABEL = "vault-autounseal-token"
+
+    def __init__(
+        self,
+        charm: CharmBase,
```

Comment:
This is not used anymore

Reply:
This is still used by the `juju_facade`, but you're right we can get rid of the instance variable.